src/verify/mod.rs

```diff
@@ -0,0 +1,309 @@
+}
+
+#[derive(Debug, Clone)]
+struct ScopeItem<'a> {
+    name: &'a str,
+    variable: Variable,
+    source: Option<ast::Node<'a>>,
+}
+
+impl<'a, 'b: 'a> TopLevelScope<'a> {
+    /// Takes a slice of top-level items and builds a `TopLevelScope` from them.
+    /// It does not verify the items as this should be done after the scope is fully constructed.
+    fn build(items: &'b [ast::Item<'b>]) -> (TopLevelScope<'a>, Vec<Error<'b>>) {
+        let mut out = TopLevelScope {
+            map: HashMap::with_capacity(items.len()),
+        };
+        let mut errors = Vec::new();
+
+        for item in items {
+            match &item.kind {
+                ast::ItemKind::FnDecl { name, params, body: _ } => {
+                    if let Some(conflict) = out.map.insert(
+                        name.name,
+                        TopLevelItem {
+                            func: Func {
+                                n_params: params.len(),
+                            },
+                            source: item,
+                        },
+                    ) {
+                        errors.push(Error {
+                            kind: error::Kind::ItemConflict(conflict.source, item),
+                            context: error::Context::TopLevel,
+                            source: item.node(),
+                        });
+                    }
+                }
+            }
+        }
+
+        (out, errors)
+    }
+
+    fn check_item(&'a self, item: &'b ast::ItemKind<'b>) -> Vec<Error<'b>> {
+        match item {
+            ast::ItemKind::FnDecl { name: _, params, body } => Scope::check_fn(self, params, body),
+        }
+    }
+
+    fn check_items(&'a self, items: &'b [ast::Item<'b>]) -> Vec<Error<'b>> {
+        let mut errors = Vec::new();
+        for item in items {
+            errors.extend(self.check_item(&item.kind));
+        }
+        errors
+    }
+}
+
+impl<'a, 'b: 'a> Scope<'a> {
+    /// Creates a new scope containing the given item with the given parent.
+    /// The TopLevelScope is inherited from `parent`.
+    fn new(item: ScopeItem<'a>, parent: &'a Scope<'a>) -> Scope<'a> {
+        Scope {
+            item: Some(item),
+            parent: Some(parent),
+            top_level: parent.top_level,
+        }
+    }
+    /// Creates a new empty scope - with no item and no parent.
+    fn empty(top_level: &'a TopLevelScope<'a>) -> Scope<'a> {
+        Scope {
+            item: None,
+            parent: None,
+            top_level,
+        }
+    }
+
+    /// Creates a new scope, a child of `self`, containing `item`.
+    fn child(&'a self, item: ScopeItem<'a>) -> Scope<'a> {
+        Self::new(item, self)
+    }
+
+    /// Given a TopLevelScope, check_fn checks the body of the function and returns any errors.
+    fn check_fn(
+        top_level: &'a TopLevelScope<'a>,
+        params: &'b ast::FnParams<'b>,
+        block: &'b ast::Block<'b>,
+    ) -> Vec<Error<'b>> {
```

Comment:
This should be a method on `TopLevelScope`, not an associated function of `Scope`

Reply:
I've moved it now, however my reason for putting it within Scope is that it uses Scopes internals but doesn't really use TopLevelScopes internals.

Reply:
Fair enough - this probably comes down to style... I would say that functions of this singature have an "obvious enough" method receiver that it would be cleaner to write `top_scope.check_fn(..)` than `Scope::check_fn(top_scope, ..)`